scrapy/spiders/__init__.py

```diff
@@ -3,21 +3,26 @@
 
 See documentation in docs/topics/spiders.rst
 """
+from __future__ import annotations
+
 import logging
-from typing import Optional
+from typing import TYPE_CHECKING, Optional
 
 from scrapy import signals
 from scrapy.http import Request
 from scrapy.utils.trackref import object_ref
 from scrapy.utils.url import url_is_from_spider
 
+if TYPE_CHECKING:
+    from scrapy.crawler import Crawler
+
 
 class Spider(object_ref):
     """Base class for scrapy spiders. All spiders must inherit from this
     class.
     """
 
-    name: Optional[str] = None
+    name: str
```

Comment:
This is a change in behavior. It looks fine, but I'm not sure I understand the consequences fully. Would it make sense to keep it as it was, or should we investigate further?

Reply:
This indeed looks wrong as base spider classes usually don't have the name attribute set.

Reply:
It's required not to be `None` by `httpcache.py`.
```
scrapy\extensions\httpcache.py:335: error: Argument 2 to "Path" has incompatible type "Optional[str]"; expected "Union[str, PathLike[str]]"  [arg-type]
```

Reply:
I'm not sure what to do about this sort of issue.

Reply:
Well.
The base spiders I mentioned aren't supposed to be run so middlewares won't be executed :) Maybe it's enough to add a cast/assert to the middleware.
OTOH there is a WIP PR #4327 that would make the name optional even for spiders that can be run.

Reply:
Isn't it required that there is a name, even for the base `Spider`s? `InitSpider` does not have a name, but when you construct it you must provide a name as a parameter or else `Spider`'s `__init__` will raise an exception. It's these lines being removed in the other PR: https://github.com/scrapy/scrapy/pull/4327/files#diff-bfeaff1ed9c7b0a9a61d76bca0e46e4e34289df77cbc1a5143c31e4d1be6fcadL28-L29

Reply:
The `Spider` class itself does not have a `name` defined, and that is not an error. And subclasses do not need a name either, as long as they are not instantiated (which is when `Spider.__init__` gets called, though a subclass could actually override its `__init__` to prevent that exception in the first place).

Reply:
If a subclass doesn't call the base `__init__` that seems like all promises made by the base class are off and functions provided by the base class may not work. Python doesn't enforce type hints at runtime so the code will work the same as before even if somebody skips over the code that asserts there is always a name.

Reply:
I'm thinking this is correct as `str`. `name` is intended not to be `None`. If the type checker is really smart, having it declared to be `str` could generate a warning if somebody does create a special subclass that doesn't call the base `__init__` to set it (I haven't checked this, but I doubt it). The type checker will definitely generate a warning if somebody writes code that explicitly sets it to `None`.